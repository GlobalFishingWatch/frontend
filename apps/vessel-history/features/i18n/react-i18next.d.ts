// import the original type declarations
// import all namespaces (for the default language, only)
import type { datasets, flags } from '@globalfishingwatch/i18n-labels'

import 'react-i18next'

import type translations from '../../public/locales/source/translations.json'

declare module 'react-i18next' {
  // and extend them!
  interface Resources {
    translations: typeof translations
    datasets: typeof datasets
    flags: typeof flags
  }
}
