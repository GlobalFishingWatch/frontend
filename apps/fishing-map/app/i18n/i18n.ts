import { createInstance } from 'i18next'
import resourcesToBackend from 'i18next-resources-to-backend'
import { initReactI18next } from 'react-i18next/initReactI18next'
import { Locale } from '@globalfishingwatch/api-types'
import { fallbackLng, getI18nConfig } from './config'

const initI18next = async (lng, ns) => {
  const i18nInstance = createInstance()
  await i18nInstance
    .use(initReactI18next)
    .use(
      resourcesToBackend((language, namespace) => import(`./locales/${language}/${namespace}.json`))
    )
    .init(getI18nConfig(lng, ns))
  return i18nInstance
}

export type useTranslationsOptions = {
  keyPrefix: string
}
export async function useTranslation(
  lng = fallbackLng as Locale,
  ns?: string | string[],
  options = {} as useTranslationsOptions
) {
  const i18nextInstance = await initI18next(lng, ns)
  return {
    t: i18nextInstance.getFixedT(lng, Array.isArray(ns) ? ns[0] : ns, options.keyPrefix),
    i18n: i18nextInstance,
  }
}
