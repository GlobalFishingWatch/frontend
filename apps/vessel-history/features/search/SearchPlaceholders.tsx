import { Trans,useTranslation } from 'react-i18next'
import cx from 'classnames'

import vesselImage from 'assets/images/vessel@2x.png'
import vesselNoResultsImage from 'assets/images/vessel-side@2x.png'

import type { HttpError } from './search.slice'

import styles from './SearchPlaceholders.module.css'

type SearchPlaceholderProps = {
  className?: string
  children?: React.ReactNode
}

type SearchErrorProps = {
  error: HttpError
  className?: string
}

type SearchNoResultsStateProps = SearchPlaceholderProps & {
  contactUsLink?: string
  onContactUsClick?: () => void
}

function SearchPlaceholder({ children, className = '' }: SearchPlaceholderProps) {
  return (
    <div className={cx(styles.emptyState, className)}>
      <div>{children}</div>
    </div>
  )
}

export function SearchNoResultsState({
  className = '',
  contactUsLink = '',
  onContactUsClick = () => { },
}: SearchNoResultsStateProps) {
  return (
    <SearchPlaceholder className={className}>
      <img src={vesselNoResultsImage.src} alt="vessel" className={styles.noResultsImage} />
      <p>
        <Trans i18nKey="search.noResults">
          Can't find the vessel you are looking for? Try using MMSI, IMO or Call Sign or{' '}
          <a
            href={contactUsLink}
            rel="noopener noreferrer"
            target="_blank"
            onClick={onContactUsClick}
          >
            contact us
          </a>{' '}
          if you still can’t find it.
        </Trans>
      </p>
    </SearchPlaceholder>
  )
}

export function SearchNoResultsFromTmtState({
  className = '',
}: SearchNoResultsStateProps) {
  return (
    <SearchPlaceholder className={className}>
      <img src={vesselNoResultsImage.src} alt="vessel" className={styles.noResultsImage} />
      <p>
        <Trans i18nKey="search.noTmtResults">
          A vessel with these criteria could not be found, due to a timeout in the search function. Please try again.
        </Trans>
      </p>
    </SearchPlaceholder>
  )
}

export function SearchEmptyState({ className = '' }: SearchPlaceholderProps) {
  const { t } = useTranslation()
  return (
    <SearchPlaceholder className={className}>
      <img src={vesselImage.src} alt="vessel" className={styles.vesselImage} />
      <p>
        {t(
          'search.description',
          'Search by vessel name or identification code (IMO, MMSI, VMS ID, etc…). You can narrow your search pressing the filter icon in the top bar'
        )}
      </p>
    </SearchPlaceholder>
  )
}

export function SearchNotAllowed({ className = '' }: SearchPlaceholderProps) {
  const { t } = useTranslation()
  return (
    <SearchPlaceholder className={className}>
      <p>{t('search.notAllowed', 'Search not allowed')}</p>
    </SearchPlaceholder>
  )
}

export function SearchErrorState({ className = '', error }: SearchErrorProps) {
  const { t } = useTranslation()
  return (
    <SearchPlaceholder className={className}>
      <p>{error.message}</p>
    </SearchPlaceholder>
  )
}

export default SearchPlaceholder
