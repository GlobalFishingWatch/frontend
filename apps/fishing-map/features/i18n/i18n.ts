import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import Backend from 'i18next-http-backend'

import { IS_DEVELOPMENT_ENV, PATH_BASENAME } from 'data/config'
import { WORKSPACE_ENV } from 'data/workspaces'
import { Locale } from 'types'

export const CROWDIN_IN_CONTEXT_LANG = 'val'

export const LocaleLabels = [
  { id: Locale.en, label: 'English' },
  { id: Locale.es, label: 'Español' },
  { id: Locale.fr, label: 'Français' },
  { id: Locale.id, label: 'Bahasa Indonesia' },
  { id: Locale.pt, label: 'Portuguese' },
]

const NPM_SCOPE = WORKSPACE_ENV === 'production' ? 'stable' : 'latest'
const SHARED_LABELS_PATH = IS_DEVELOPMENT_ENV
  ? 'http://localhost:8000'
  : `https://cdn.jsdelivr.net/npm/@globalfishingwatch/i18n-labels@${NPM_SCOPE}`

const PACKAGE_NAMESPACES = ['flags', 'datasets', 'timebar']

i18n
  // load translation using http -> see /public/locales
  // learn more: https://github.com/i18next/i18next-http-backend
  .use(Backend)
  // detect user language
  // learn more: https://github.com/i18next/i18next-browser-languageDetector
  .use(LanguageDetector)
  // pass the i18n instance to react-i18next.
  .use(initReactI18next)
  // init i18next
  // for all options read: https://www.i18next.com/overview/configuration-options
  .init({
    backend: {
      loadPath: (lngs: string[], namespaces: string[]) => {
        if (namespaces.some((namespace: string) => PACKAGE_NAMESPACES.includes(namespace))) {
          return `${SHARED_LABELS_PATH}/{{lng}}/{{ns}}.json`
        }
        return `${PATH_BASENAME}/locales/{{lng}}/{{ns}}.json`
      },
    },
    ns: ['translations', 'flags', 'datasets', 'timebar'],
    defaultNS: 'translations',
    fallbackLng: Locale.en,
    supportedLngs: [...Object.values(Locale), CROWDIN_IN_CONTEXT_LANG],
    debug: process.env.i18n_DEBUG === 'true',
    interpolation: {
      escapeValue: false, // not needed for react as it escapes by default,
      // format: (value, format, lng) => {
      //   if (format === 'intlDate') {
      //     return new Intl.DateTimeFormat(lng).format(value)
      //   }
      //   return value
      // },
    },
    react: {
      // wait: true,
      useSuspense: false,
    },
  })

i18n.on('languageChanged', (lng) => {
  if (process.browser) {
    document.documentElement.setAttribute('lang', lng)
  }
})

const t = i18n.t.bind(i18n)

export { t }

export default i18n
