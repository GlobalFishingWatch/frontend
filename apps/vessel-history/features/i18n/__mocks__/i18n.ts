import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'

i18n.use(initReactI18next).init({
  react: { useSuspense: false },
  lng: 'en',
  fallbackLng: 'en',

  // have a common namespace used around the full app
  ns: ['translations', 'flags', 'datasets'],
  defaultNS: 'translations',

  debug: process.env.i18n_DEBUG === 'true',

  interpolation: {
    escapeValue: false, // not needed for react!!
  },

  resources: { en: { translations: {} } },
})

export default i18n
