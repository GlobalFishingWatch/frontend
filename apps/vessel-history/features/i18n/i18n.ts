import { initReactI18next } from 'react-i18next'
import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import Backend from 'i18next-http-backend'

import { WORKSPACE_ENV } from 'data/config'
import { PATH_BASENAME } from 'routes/routes'
import { Locale } from 'types'

export const LocaleLabels = [
  { id: Locale.en, label: 'English' },
  // { id: Locale.es, label: 'Español' },
  { id: Locale.fr, label: 'Français' },
]

const NPM_SCOPE = WORKSPACE_ENV === 'production' ? 'stable' : 'latest'
export const SHARED_LABELS_PATH =
  process.env.NODE_ENV === 'development'
    ? 'http://localhost:8000'
    : `https://cdn.jsdelivr.net/npm/@globalfishingwatch/i18n-labels@${NPM_SCOPE}`

export const PACKAGE_NAMESPACES = ['flags', 'datasets']

i18n
  // load translation using http -> see /public/locales
  // learn more: https://github.com/i18next/i18next-http-backend
  .use(Backend)
  // detect user language
  // learn more: https://github.com/i18next/i18next-browser-languageDetector
  .use(LanguageDetector)
  // pass the i18n instance to react-i18next.
  .use(initReactI18next)
  // init i18next
  // for all options read: https://www.i18next.com/overview/configuration-options
  .init({
    backend: {
      loadPath: (lngs: string[], namespaces: string[]) => {
        if (namespaces.some((namespace: string) => PACKAGE_NAMESPACES.includes(namespace))) {
          return `${SHARED_LABELS_PATH}/{{lng}}/{{ns}}.json`
        }
        return `${PATH_BASENAME}/locales/{{lng}}/{{ns}}.json`
      },
    },
    ns: ['translations', 'flags', 'datasets'],
    defaultNS: 'translations',
    fallbackLng: Locale.en,
    supportedLngs: Object.values(Locale),
    debug: process.env.i18n_DEBUG === 'true',
    interpolation: {
      escapeValue: false, // not needed for react as it escapes by default,
      // format: (value, format, lng) => {
      //   if (format === 'intlDate') {
      //     return new Intl.DateTimeFormat(lng).format(value)
      //   }
      //   return value
      // },
    },
  })

i18n.on('languageChanged', (lng) => {
  if (process.browser) {
    document.documentElement.setAttribute('lang', lng)
  }
})

const t = i18n.t.bind(i18n)

export { t }

export default i18n
