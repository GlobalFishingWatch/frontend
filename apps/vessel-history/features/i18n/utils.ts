import type { MultiSelectOption } from '@globalfishingwatch/ui-components'

import { t } from './i18n'

export const getPlaceholderBySelections = (selections?: MultiSelectOption[]): string => {
  if (!selections?.length) return t('selects.allSelected', 'All')
  return selections.length > 1
    ? `${selections.length} ${t('selects.selected', 'selected')}`
    : selections[0].label.toString()
}

export const removeDatasetVersion = (datasetId: string) => {
  return datasetId ? datasetId?.split(':')[0] : ''
}

export const getDatasetNameTranslated = (dataset: { id: string; name?: string }): string => {
  return t(`datasets:${removeDatasetVersion(dataset.id)}.name`, dataset.name)
}
export const getDatasetDescriptionTranslated = (dataset: {
  id: string
  description?: string
}): string => {
  return t(`datasets:${removeDatasetVersion(dataset.id)}.description`, dataset.description)
}
