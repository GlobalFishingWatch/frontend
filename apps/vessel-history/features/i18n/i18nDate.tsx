import React, { Fragment } from 'react'
import { useTranslation } from 'react-i18next'
import type { DateTimeFormatOptions} from 'luxon';
import { DateTime, Zone } from 'luxon'
import type { Locale } from 'types'

import { getUTCDateTime } from 'utils/dates'

import i18n from './i18n'

type Dates = {
  date: string | number
  format?: DateTimeFormatOptions
}

type TMTDate = {
  date: string | number
  format?: DateTimeFormatOptions
  forcedTokensFormat?: string
}

type formatI18DateParams = {
  format?: DateTimeFormatOptions
  locale?: Locale
  tokensFormat?: string
}

export const formatI18nDate = (
  date: string | number,
  {
    format = DateTime.DATE_MED,
    locale = i18n.language as Locale,
    tokensFormat,
  }: formatI18DateParams = {}
) => {
  const dateTimeDate = getUTCDateTime(date).setLocale(locale)
  // console.log(
  //   (() => {
  //     const a = DateTime.fromISO('20200101').setZone('utc', { keepLocalTime: true })
  //     const b = DateTime.fromISO('2020-01-01T00:00:00', { zone: 'utc' })
  //     const c = DateTime.fromISO('2020-01-01T00:00:00')
  //     return {
  //       a: a.toISO(),
  //       b: b.toISO(),
  //       c: c.toUTC().toISO(),
  //     }
  //   })()
  // )
  // console.log(new Date().getTime() - DateTime.now().toUTC().toMillis())
  const formattedDate = tokensFormat
    ? dateTimeDate.toFormat(tokensFormat)
    : dateTimeDate.toLocaleString(format)
  return `${formattedDate}${format === DateTime.DATETIME_MED ? ' UTC' : ''}`
}

export const useI18nDate = (
  date: string | number,
  format = DateTime.DATE_MED,
  tokensFormat?: string
) => {
  const { i18n } = useTranslation()
  return formatI18nDate(date, { format, locale: i18n.language as Locale, tokensFormat })
}

const I18nDate = ({ date, format = DateTime.DATE_MED }: Dates) => {
  const dateFormatted = useI18nDate(date, format)
  return <Fragment>{dateFormatted}</Fragment>
}

export const formatI18nSpecialDate = ({ date, format = DateTime.DATE_MED }: TMTDate) => {
  let parsedDate, tokensFormat
  if (date.toString().match(/^\d{4}(00|99){1,2}$/)) {
    parsedDate = date.toString().replace(/(\d{4})(00|99){1,2}$/, '$1-01-01')
    tokensFormat = 'yyyy'
  } else if (date.toString().match(/^\d{6}(00|99)$/)) {
    parsedDate = date.toString().replace(/(\d{4})(\d{2})(00|99)$/, '$1-$2-01')
    tokensFormat = 'LLL, yyyy'
  } else if (date.toString().match(/^\d{4}-\d{2}-9{2}$/)) {
    // Converts 2016-02-99 dates to 2016-02-01 and formats it as Feb, 2016
    // TODO: Confirm with the source if this interpretation is correct
    parsedDate = date.toString().replace(/(\d{4})-(\d{2})-(9{2})$/, '$1-$2-01')
    tokensFormat = 'LLL, yyyy'
  } else if (date.toString().match(/^\d{8}$/)) {
    parsedDate = date.toString().replace(/(\d{4})(\d{2})(\d{2})$/, '$1-$2-$3')
  }
  return formatI18nDate(parsedDate ?? date, {
    format,
    locale: i18n.language as Locale,
    tokensFormat: tokensFormat,
  })
}

export const I18nSpecialDate = ({ date, format = DateTime.DATE_MED }: TMTDate) => {
  const dateFormatted = formatI18nSpecialDate({ date, format })
  return <Fragment>{dateFormatted}</Fragment>
}

export default I18nDate
