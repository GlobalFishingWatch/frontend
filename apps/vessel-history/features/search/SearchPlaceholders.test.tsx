import { Fragment } from 'react'
import { I18nextProvider } from 'react-i18next'
import { render } from '@testing-library/react'

import i18n from 'features/i18n/i18n'

import { SearchEmptyState, SearchNoResultsState, SearchNotAllowed } from './SearchPlaceholders'

jest.mock('features/i18n/i18n')
// This is to setup i18n provider for the test
render(
  <I18nextProvider i18n={i18n}>
    <Fragment />
  </I18nextProvider>
)

describe('<SearchNoResultsState />', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('shows SearchNoResultsState', () => {
    const component = render(<SearchNoResultsState className="foobar" />)
    expect(component.asFragment()).toMatchSnapshot()
  })
})

describe('<SearchEmptyState />', () => {
  it('shows SearchEmptyState', () => {
    const component = render(<SearchEmptyState className="foobar" />)
    expect(component.asFragment()).toMatchSnapshot()
  })
})

describe('<SearchNotAllowed />', () => {
  it('shows SearchNotAllowed', () => {
    const component = render(<SearchNotAllowed className="foobar" />)
    expect(component.asFragment()).toMatchSnapshot()
  })
})
