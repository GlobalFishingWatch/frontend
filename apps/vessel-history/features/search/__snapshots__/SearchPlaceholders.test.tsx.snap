// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchEmptyState /> shows SearchEmptyState 1`] = `
<DocumentFragment>
  <div
    class="emptyState foobar"
  >
    <div>
      <img
        alt="vessel"
        class="vesselImage"
      />
      <p>
        Search by vessel name or identification code (IMO, MMSI, VMS ID, etc…). You can narrow your search pressing the filter icon in the top bar
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SearchNoResultsState /> shows SearchNoResultsState 1`] = `
<DocumentFragment>
  <div
    class="emptyState foobar"
  >
    <div>
      <img
        alt="vessel"
        class="noResultsImage"
      />
      <p>
        Can't find the vessel you are looking for? Try using MMSI, IMO or Call Sign or 
        <a
          href=""
          rel="noopener noreferrer"
          target="_blank"
        >
          contact us
        </a>
         if you still can’t find it.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SearchNotAllowed /> shows SearchNotAllowed 1`] = `
<DocumentFragment>
  <div
    class="emptyState foobar"
  >
    <div>
      <p>
        Search not allowed
      </p>
    </div>
  </div>
</DocumentFragment>
`;
